src/sys.zig

```diff
@@ -443,6 +443,90 @@ pub fn Maybe(comptime ReturnTypeT: type) type {
     return JSC.Node.Maybe(ReturnTypeT, Error);
 }
 
+/// Returns `true` if the two directories are on the same filesystem
+///
+/// This function does this test by trying to rename a file named from `dir_with_file_to_rename` -> `dir_to_rename_file_to`
+///
+/// The file must have the name `tmpname`
+pub fn testSameFileSystem(
+    dir_with_file_to_rename: std.fs.Dir,
+    dir_to_rename_file_to: std.fs.Dir,
+    tmpname: [:0]const u8,
+) bool {
+    // Make sure cachedir and cwd are in the same filesystem
+    std.posix.renameatZ(dir_with_file_to_rename.fd, tmpname, dir_to_rename_file_to.fd, tmpname) catch return false;
+    return true;
+}
+
+/// Variant of `testSameFileSystem` except the file to rename to is the abspath
+pub fn testSameFileSystemPaths(
+    dir_with_file_to_rename: std.fs.Dir,
+    abspath: [:0]const u8,
+    tmpname: [:0]const u8,
+) bool {
+    std.posix.renameatZ(dir_with_file_to_rename.fd, tmpname, std.fs.cwd().fd, abspath) catch return false;
+    return true;
+}
+
+pub const PosixPathComponentIter = struct {
+    path_to_iterate: []const u8,
+    start: usize = 0,
+
+    pub fn next(this: *PosixPathComponentIter) ?[]const u8 {
+        if (this.start >= this.path_to_iterate.len) return null;
+        const slash_idx = this.start + (std.mem.indexOfScalar(u8, this.path_to_iterate[this.start..], '/') orelse {
+            const remaining = this.path_to_iterate[this.start..];
+            if (remaining.len == 0) return null;
+            this.start = this.path_to_iterate.len;
+            return remaining;
+        });
+        if (slash_idx == 0) {
+            this.start = 1;
+            return "/";
+        }
+        this.start = slash_idx + 1;
+        return this.path_to_iterate[0..slash_idx];
+    }
+};
+
+/// Invariants:
+///
+/// - node_modules_folder_path is absolute path with no relative syntax
+pub fn findBestDirectoryInSameFileSystem(
```

Comment:
This can go in the package manager code

we don't need to do this elsewhere afaict